### ISSUE TITLE
The future of Object.merge

### ISSUE THREAD

**Issue by andrewplummer** [andrewplummer/Sugar#518]
`Object.merge` has had a bit of a complicated history. Although it seems straightforward enough to merge properties from one object into another, there are a number of different ways to do it, and things that can only be handled during the merge operation itself. `Date.create` has had a similar problem, and beginning in 2.0.0 an options object will handle some of the smaller edge cases when creating dates. I'm not a huge fan of options objects as it's hard to remember the names of options -- having them all over Sugar would be horrible. However, once you have enough arguments it becomes a similar problem to remember their order, and it's also a hassle to write out `(false, false, true)` just to get to the option you want. `Object.merge` is becoming like this. There are 2 options now ("deep" and "resolve") and I would like to add 2 more. I'll list them out here (abbreviating here and there for space):
#### Deep Merging

This should be pretty obvious -- just recursively merging if properties themselves are objects. This is currently the 3rd argument and is `false` by default. Proposal is to stay false by default and become `deep` on the options object.

```
var o1 = { name: { first: 'Darth' } };
var o2 = { name: { last: 'Vader' } };
merge(o1, o2); ->  { name: { last: 'Vader' } }
merge(o1, o2, true); ->  { name: { first: 'Darth', last: 'Vader' } }
merge(o1, o2, { deep: true }); // with options
```
#### Resolve

When a conflict arises in a merge, which property should win? Right now the default is `true`, which means the source object's property wins -- in other words what you would expect from a standard merge. If this argument is `false` then the target's property wins, or what you would expect from a `defaults` method (side proposal: create a `defaults` method as a shortcut to this behavior). However this argument can also be passed a function:

```
var o1 = { name: 'Darth'  };
var o2 = { name: 'Vader' };
var fn = function (key, targetVal, sourceVal) {
  return targetVal + ' ' + sourceVal;
}
merge(o1, o2); ->  { name: 'Vader' }
merge(o1, o2, false, false); ->  { name: 'Darth' }
merge(o1, o2, false, fn); ->  { name: 'Darth Vader' }
merge(o1, o2, { resolve: fn }); // with options
```

The resolver function is intended to allow fine grained control over merge conflicts. However, right now it has some confusing behavior that can almost be considered buggy. First, deep merges will continue to step in to an object returned by the resolver function. This is confusing because the sprit of the resolver function is to allow the user to override standard merge behavior, so anything returned by the function should become the final value. Conversely, right now the resolver function will _always_ use the return value of the function, even if that is undefined. I would like to change this so that the final value of the resolver function be used _if defined_. If undefined, then `Object.merge` will handle the merge in the default way. This will make it more intuitive for users to be able to step in and take control of merges for certain properties only, make it more aligned with the rest of `Object.merge` (as undefined properties are not merged normally), and also be in line with Lodash (not top priority, but nice). So (I'm squashing both behaviors here into one example):

Current Behavior:

```
var o1 = { user_count: 5, foo: 'bar', obj: { inner: 1 } };
var o2 = { user_count: 5, foo: 'baz', obj: { inner: 2 } };
var fn = function (key, targetVal, sourceVal) {
  if (key === 'user_count') {
    return targetVal + sourceVal;
  }  else if (key === 'obj') {
    // New object, something else entirely
    return { a: 'b' }
  }
}
// "foo" wasn't handled so it becomes undefined
// "obj" was passed a new object, but since it is a deep merge the 
// result is traversed into. Since "inner" isn't handled it becomes undefined.
merge(o1, o2, true, fn); ->  { user_count: 10, foo: undefined, obj: { inner: undefined } }
```

Proposed behavior:

```
// "foo" wasn't handled so it gets handled in the default way
// "obj" was passed a new object, so this is respected and becomes the final value
merge(o1, o2, { deep: true, resolve: fn }); ->  { user_count: 10, foo: 'baz', obj: { a: 'b' } }
```
#### Non-enumerable Properties

`Object.merge` currently only operates on enumerable properties. This should be correct for most use cases, however there are times when merging non-enumerable properties is desirable as well. Proposal is to add this as an option.

Current behavior:

```
var o1 = { foo: 'bar' };
Object.defineProperty(obj, 'hiddenFoo', {
  value: 'hiddenBar',
  enumerable: false
});
var o2 = merge({}, o1);
console.log(o1.foo); // 'bar'
console.log(o1.hiddenFoo; // 'hiddenBar'
console.log(o2.foo); // 'bar'
console.log(o2.hiddenFoo; // undefined
```

Proposed behavior:

```
var o2 = merge({}, o1, { hidden: true });
console.log(o1.foo); // 'bar'
console.log(o1.hiddenFoo; // 'hiddenBar'
console.log(o2.foo); // 'bar'
console.log(o2.hiddenFoo; // 'hiddenBar'
```

As for the naming, I am proposing `hidden` because `non-enumerable` is difficult to write out, and `enumerable: false` does not mean the same thing. If there are other ideas for naming that would make this clearer, I'm open to them.
#### Merging Property Descriptors

Currently, only properties with an associated "value" are accessed and merged. As with non-enumerable properties, I think that this is fine as a default. However, ES5 introduced accessors in the form of property descriptors, and I think it's time that Sugar finally supported this as an option.

Current behavior:

```
var data = { label: 'foo' };
var o1 = {
  get label() {
    return data.label;
  },
  set label(value) {
    data.label = value;
  }
}
var o2 = merge({}, o1);
console.log(o1.label, o2.label);  // foo foo 
data.label = 'bar';
console.log(o1.label, o2.label);  // bar foo (link created by setter is broken)
```

Proposed behavior:

```
merge({}, o2, { descriptor: true });  // produces a faithful merge of all accessors
```

As for naming `descriptor` seems to make sense, although I'm not sure about singular vs. plural. Also open to other naming options as well. Note that property descriptors are not supported in IE8 and below, so this option will do nothing in those environments. However, accessors can't be used in the first place, so if you need to support <=IE8 you won't need this option anyway.
### Summary

To summarize the proposals here:
- Using an options object instead of listed arguments
- Adding an option for non-enumerable properties (off by default)
- Adding an option for property descriptors (off by default)
- Changing the resolve function to not go deep (respect the final value if returned)
- Changing the resolve function to do a default merge if the final value is undefined
- Adding a method called `Object.defaults` that is a shortcut to `resolve: false`.

**Comment by ermouth** [comment]
Read carefully, looks really great. Both flexible and backward-compatible. Thanks!

**Comment by andrewplummer** [comment]
So glad you're happy... I know you've been through a lot :D
What do you think about the `Object.defaults` idea? Would you use it?

**Comment by ermouth** [comment]
> What do you think about the Object.defaults idea? Would you use it?

I understand why did you name it .defaults, but it would be bit confusing for those who will _read_  code. Is there any short word in English, that means filling in the gaps ot smth like this? Maybe just `Object.fill`?

**Comment by andrewplummer** [comment]
> I understand why did you name it .defaults, but it would be bit confusing for those who will read code.

Well that's interesting. I'm usually the one going for more readable naming. However recently I feel I've been too haphazard about it. I'm all for readable, but if there is a name out there that really is more commonly accepted and guessable, then I think it should win. I feel that `defaults` is like that, but I may be wrong. Looking at it I can mostly only find Underscore/Lodash that uses it.

But hold on a second. Actually I also wanted to bring up another idea. Merging multiple objects together is also a common operation, so I was thinking that there also may be a need for `Object.mergeAll` or possibly another option on `merge` telling it to combine all array objects together.

1. What would you think of a `mergeAll`?
2. If we added that, it would probably remove the need for `defaults`. Ordering the objects a certain way would decide which one wins (this is more jQuery style)

**Comment by ermouth** [comment]
> I'm usually the one going for more readable naming

Indeed, Sugar naming is very nice (`.has` is especially cute :). But as for `.defaults` – it‘s not only my opinion, that‘s why I‘ve taken a little pause. I‘ve asked 3 guys about it, showing em code. Only one gave partially valid answer. It‘s quite weak argument, but it shows realities.

As for me, `.mergeAll` looks definitely much more clear, despite of camelCase.

**Comment by andrewplummer** [comment]
> Only one gave partially valid answer. It‘s quite weak argument, but it shows realities.

No, I think it's great.. I especially want real opinions. But are you saying that only one liked `defaults`, and two didn't like it?

> As for me, .mergeAll looks definitely much more clear, despite of camelCase.

I agree... and I also agree about the camelCase, like to keep it one word when possible but unfortunately it's not always possible.

So here's another question (I never stop) ... so if you had `Object.mergeAll([a,b,c,d,e])` then should it merge everything into `a` or into a new object?

**Comment by ermouth** [comment]
Seems `.mergeAll` is an array method, not object. So `[a,b,c].mergeAll()`, not `Object.mergeAll([a,b,c])`. As for me, `Object.mergeAll()` should take plain objects as arguments.

> should it merge everything into `a`

Yes, first element acts as accumulator. If one wants new object, let him write `[{}, b, c].mergeAll();`

**Comment by andrewplummer** [comment]
> So [a,b,c].mergeAll(), not Object.mergeAll([a,b,c])

Ah, well I wasn't expecting that. That's definitely not a direction I'd like to take this. Merging objects together is something that should be a responsibility of the `Object` class/package and I definitely want to keep it there for various reasons. Also, I don't think it makes sense for a method to become an array method just because it accepts arrays.

> Yes, first element acts as accumulator. If one wants new object, let him write [{}, b, c].mergeAll();

I tend to agree with this though.

**Comment by ermouth** [comment]
> I don't think it makes sense for a method to become an array method just because it accepts arrays

Can‘t understand this, it _surely is_ array method ) You don‘t write Object.map() or Object.reduce() for arrays – and`.mergeAll` is, actually, kinda reduce.

**Comment by andrewplummer** [comment]
> it surely is array method

Well it is in the sense that it *could* be mapped onto an array, but those other methods are performing operations on arrays, whereas this is all about performing operations on objects, which is the main difference. Of course you're right that `reduce` is a bit different, but it's kind of the odd-man on the Array class and it isn't doing anything more important than a series of callbacks with an accumulator. `Object.mergeAll` is doing merge operations on the objects themselves. The array is completely unimportant to the implementation, it's just a holder. For example it could take enumerated arguments just as easily (`Object.mergeAll(a, b, c)`), it makes no difference, really.

For another example, `Date.UTC` accepts a list of parameters but that doesn't mean you would make `Array#createDate` simply because it's passed a list. The point is about operating on dates.

**Comment by ermouth** [comment]
`.mergeAll([obj1, obj2, objn])` is definitely fold of a list, that is denoted as explicit array – by proposed syntax.

If you proposed `Object.mergeAll( obj1, obj2, objn )` – I would never say it‘s an array method, despite it‘s also fold.

`Date.UTC` example is not relevent in this case – it does not take array, it takes list of args.

Although, since Sugar is your lib, and I like it as is – I‘m stopping here, just to avoid ping-pong. Do what you think is reasonable )

**Comment by andrewplummer** [comment]
For me, list of args or array are mostly an implementation detail and in any case not relevant to the question of where a method should exist. `Object.mergeAll`, if it's added at all, will definitely exist on `Object` because it is all about merging objects together. This is for both because it's conceptually more obvious, but also for technical reasons as the code required for this lives in the object package and it can't exist in both.

**Comment by Slayer95** [comment]
Suggestion for ``resolver`` function:

```
var fn = function (key, targetVal, sourceVal, undefined) { // note the extra parameter
  if (key === 'user_count') {
    return targetVal + sourceVal;
  }  else if (key === 'obj') {
    // New object, something else entirely
    return { a: 'b' }
  } else if (key === 'obj2') {
  return undefined; // assigns the value ``undefined``
  }
//return void 0; // ignores the property; statement may be left out
}
```

Internally, the resolver function could be called with a 4th parameter, a private ``sugar`` object, so that if the result of the resolver function strictly equals to it, then the property value will be ``undefined``, rather than ignored. Of course, if the ``Object.merge`` user doesn't want this behavior, they would simply pass a function with length 3.

This would allow fine-grained control.over the behavior of ``undefined``.

**Comment by andrewplummer** [comment]
That's an interesting idea. Note that I'm also planning on adding the `target` and `source` objects to the arguments to give more control. Another possibility is that instead of `undefined`, it could be the inverse and the passed object could be a way to tell it to continue the merge, and `undefined` returned would actually be `undefined`. I like your idea better though.

However all that said, I think I'm ok with Sugar taking a hard stance that `undefined` properties are equivalent to non-existent properties, and that this use case does not have to be directly handled. `null` exists instead when something needs to be defined, but no value. If this use case absolutely has to be supported then it can be worked around by directly setting `undefined` on the `target` argument once that is passed.

**Comment by andrewplummer** [comment]
OK this one is now finally locked down. First, there will be 2 new methods, `mergeAll` and `defaults`. If `mergeAll` sees an array for `source` it simply treats it as a list of objects to be merged into `target` instead of merging properties of the array itself (`0`, `1`, etc). `defaults` is the same as `mergeAll` but sets the `resolve` option to `false`, meaning the properties of `target` will win if defined. All 3 methods are on the `Object` class and are static methods, however are also mapped onto extended objects (`Object.extended()`). Options for all 3 methods are the same. I'll let the shiny new docs describe:

```
deep:        If `true` deep properties are merged recursively.
             (Default is `false`)

resolve:     Determines which property wins in the case of conflicts.
             If `true`, <source> wins. If `false`, <target> wins. If a
             function is passed, its return value will decide the result.
             Any non-undefined return value will resolve the conflict
             for that property (will not continue if `deep`). Returning
             `undefined` will do nothing (no merge). Finally, returning
             the global object `Sugar` will allow Sugar to handle the
             merge as normal. Resolve functions are passed the arguments:
             `key`, `targetValue`, `sourceValue`, `target`, `source`.
             (Default is `true`)

hidden:      If `true`, non-enumerable properties will be merged as well.
             (Default is `false`)

descriptor:  If `true`, properties will be merged by property descriptor.
             Use this option to merge getters or setters, or to preserve
             properties like `enumerable`, `configurable`, etc.
             (Default is `false`)
```

**Comment by andrewplummer** [comment]
@Slayer95 Thanks for your input here. In the end I went with a slightly modified form of your idea. In the end the motivation was the fact that there was no way to bail out of merging a property in a resolve function as returning `undefined` meant "merge as normal". I feel that returning the global `Sugar` object is much more intuitive as it is telling the result to "let Sugar handle the merge".

**Comment by Slayer95** [comment]
@andrewplummer You are welcome! Looking forward to 2.0 :)
